src/hvac.h

```diff
@@ -15,14 +15,83 @@ float ASHPCap95FromCap47( float cap47, bool useRatio, float ratio9547);
 float ASHPCap47FromCap95( float cap95, bool useRatio, float ratio9547);     
 void ASHPConsistentCaps( float& cap95, float& cap47, bool useRatio, float ratio9547);
 
+///////////////////////////////////////////////////////////////////////////////
+// class CourierMsgHandler: Courier-derived handler for library callback
+//		 messages (used by e.g. Btwxt)
+///////////////////////////////////////////////////////////////////////////////
+
+#include <courier/courier.h>
+
+// abstract base class
+class CourierMsgHandlerBase : public Courier::Courier
+{
+public:
+	void receive_error(const std::string& crMsg) override { forward_message(MSGTY::msgtyERROR, crMsg); }
+	void receive_warning(const std::string& crMsg) override { forward_message(MSGTY::msgtyWARNING, crMsg); }
+	void receive_info(const std::string& crMsg) override { forward_message( MSGTY::msgtyINFO, crMsg); }
+	void receive_debug(const std::string& crMsg) override { forward_message(MSGTY::msgtyDEBUG, crMsg); }
+
+private:
+	virtual void forward_message(MSGTY msgty, const std::string& crMsg) = 0;
+
+};	// class CourierMsgHandlerBase
+//-----------------------------------------------------------------------------
+class CourierMsgHandler : public CourierMsgHandlerBase
+{
+public:
+	using MsgCallbackFunc = void(void* pContext, MSGTY msgty, const char* msg);
+
+	CourierMsgHandler(MsgCallbackFunc* pMsgCallbackFunc,void* context)
+		: cmh_pMsgCallbackFunc(pMsgCallbackFunc), cmh_context( context)
+	{ }
+
+private:
+	virtual void forward_message(MSGTY msgty, const std::string& crMsg) override
+	{
+		const char* msg = crMsg.c_str();
+		if (cmh_pMsgCallbackFunc)
+			(*cmh_pMsgCallbackFunc)(cmh_context, msgty, msg);
+		else
+			err(PABT, "nullptr cmh_pMsgCallbackFunc '%s'", msg);
+
+	}
+
+private:
+	void* cmh_context;						// caller context
+	MsgCallbackFunc* cmh_pMsgCallbackFunc;	// pointer to callback function
+
+};	// class CourierMsgHandler
```

Comment:
My thought is that `void* context` and a `MsgCallbackFunc*` are very "C". There is a lot of indirection for people trying to read the code. My preference would be to make a new class for each different context with a specific `forward_message` override.

Reply:
1. Don't want 10 essentially identical implementations.  Maybe there is a way to template this to make it clearer?  In any case, there is going to be a "callable" somewhere aka a pointer.
2. Additionally, what is wrong with C?  CSE is small and fast, partially due to using lots of C stuff.
3. I recently learned about a "dependency injection" scheme which might be very useful here.